diff --git a/build/react-navigation/stack/views/Stack/StackView.js b/build/react-navigation/stack/views/Stack/StackView.js
index 098bd17f31f97184d227d1ce1514cb4632509de8..895bb1727b8e6a4be7cf38e61f06b4a69333cf0e 100644
--- a/build/react-navigation/stack/views/Stack/StackView.js
+++ b/build/react-navigation/stack/views/Stack/StackView.js
@@ -77,15 +77,6 @@ class StackView extends React.Component {
             if (closingRoutes.length) {
                 routes.push(...closingRoutes);
             }
-            let descriptors = props.descriptors;
-            let previousDescriptors = state.previousDescriptors;
-            if (props.descriptors !== state.previousDescriptors) {
-                descriptors = routes.reduce((acc, route) => {
-                    acc[route.key] = (props.descriptors[route.key] || state.descriptors[route.key]);
-                    return acc;
-                }, {});
-                previousDescriptors = props.descriptors;
-            }
             if (!isArrayEqual(allRoutes, previousRoutes)) {
                 // if any route objects have changed, we should update them
                 const map = allRoutes.reduce((acc, route) => {
@@ -94,11 +85,15 @@ class StackView extends React.Component {
                 }, {});
                 routes = routes.map((route) => map[route.key] || route);
             }
+            const descriptors = [...routes, ...props.state.preloadedRoutes].reduce((acc, route) => {
+                acc[route.key] = (props.descriptors[route.key] || state.descriptors[route.key]);
+                return acc;
+            }, {});
             return {
                 routes,
                 previousState: props.state,
                 descriptors,
-                previousDescriptors,
+                previousDescriptors: props.descriptors,
             };
         }
         // Here we determine which routes were added or removed to animate them
@@ -198,7 +193,7 @@ class StackView extends React.Component {
         if (!routes.length) {
             throw new Error('There should always be at least one route in the navigation state.');
         }
-        const descriptors = allRoutes.reduce((acc, route) => {
+        const descriptors = [...routes, ...props.state.preloadedRoutes].reduce((acc, route) => {
             acc[route.key] = (props.descriptors[route.key] || state.descriptors[route.key]);
             return acc;
         }, {});
